cajita/src/Cajita_FastFourierTransform.hpp

```diff
@@ -23,291 +23,509 @@
 };
 #endif
+template <class T>
+struct is_cuda_complex
+    : public is_cuda_complex_impl<typename std::remove_cv<T>::type>::type
+{
+};
 
-template <>
-struct HeffteMemoryTraits<Kokkos::HostSpace>
+template <typename>
+struct is_std_complex_impl : public std::false_type
+{
+};
+template <class Scalar>
+struct is_std_complex_impl<std::complex<Scalar>> : public std::true_type
+{
+};
+template <class T>
+struct is_std_complex
+    : public is_std_complex_impl<typename std::remove_cv<T>::type>::type
 {
-    static constexpr heffte_memory_type_t value = HEFFTE_MEM_CPU;
 };
 
 //---------------------------------------------------------------------------//
+/*!
+  \class FastFourierTransformParams
+  \brief Parameters controlling details for fast Fourier transforms.
+*/
 class FastFourierTransformParams
 {
+    bool alltoall = true;
+    bool pencils = true;
+    bool reorder = true;
+
   public:
     /*!
-      \brief Default constructor to disable aggregate initialization.
+      \brief setAllToAll Set MPI communication.
+      \param value Use all to all MPI communication.
     */
-    FastFourierTransformParams()
-        : collective( 2 )
-        , exchange( 0 )
-        , packflag( 2 )
-        , scaled( 1 )
-    {
-    }
-
+    void setAllToAll( bool value ) { alltoall = value; }
+    /*!
+      \brief setPencils Set data exchange type (pencil or slab).
+      \param value Use pencil (true) or slab (false) decomposition.
+    */
+    void setPencils( bool value ) { pencils = value; }
     /*!
-      \brief Set the collective type.
-      \param type Collective type.
+      \brief setReorder Set data handling (contiguous or strided memory).
+      \param value Use contiguous (true) or strided (false) memory layout.
+      Contiguous layout requires tensor transposition; strided layout does not.
+    */
+    void setReorder( bool value ) { reorder = value; }
+    /*!
+      \brief getAllToAll Get MPI communication.
+      \return Using all to all MPI communication or not.
+    */
+    bool getAllToAll() const { return alltoall; }
+    /*!
+      \brief getPencils Get data exchange type (pencil or slab).
+      \param value Using pencil (true) or slab (false) decomposition.
+    */
+    bool getPencils() const { return pencils; }
+    /*!
+      \brief getReorder Get data handling (contiguous or strided memory).
+      \param value Using contiguous (true) or strided (false) memory layout.
+      Contiguous layout requires tensor transposition; strided layout does not.
+    */
+    bool getReorder() const { return reorder; }
+};
 
-      0: point-to-point
-      1: all-to-all
-      2: combination
+//---------------------------------------------------------------------------//
+/*!
+  \class FastFourierTransform
+  \brief 3D distributed fast Fourier transform base implementation.
+*/
+template <class EntityType, class MeshType, class Scalar, class DeviceType,
+          class InterfaceType>
```

Comment:
Rename `InterfaceType` to `Derived`. It will make it very obvious that you implemented the CRTP.

Reply:
Done